### DISCUSSION TITLE
Making a Multi-Layered Item-Model with Datagen

### DISCUSSION THREAD

**Discussion by theoretischer** [FabricMC/fabric-api#3496] [Mod Dev Support] [answered]
### Description

Hi, i was trying to make an item Model with datagen that has multiple layers from files.
at the end i want a model like this:
{
  "parent": "item/handheld",
  "textures": {
    "layer0": "testmod:item/sticks/custom_stick",
    "layer1": "testmod:item/tool_heads/custom_tool"
  }
}

normally i generate models with
itemModelGenerator.register(EXAMPLEITEM, Models.GENERATED);

and the nearest thing i found was Models.GENERATED_TWO_LAYERS, but I have no idea how to use it and didn't find anything on the web. Help would be appreciated

**Comment by Linguardium** [comment]
based on the register method you referenced:
```
	public final void register(Item item, Item texture, Model model) {
		model.upload(ModelIds.getItemModelId(item), TextureMap.layer0(texture), this.writer);
	}
```
And looking at usages of GENERATED_TWO_LAYERS, such as the uploadArmor method:
```
	public final void uploadArmor(Identifier id, Identifier layer0, Identifier layer1) {
		Models.GENERATED_TWO_LAYERS.upload(id, TextureMap.layered(layer0, layer1), this.writer);
	}
```
it looks fairly simple to do away with the helper methods, but also appears you can just use uploadArmor directly

**Reply by theoretischer** [reply]
Hi, thanks for the reply. This is almost what I want to achieve (I didn't specify it in my original post) I want to have a Handheld item with two layers. I have tried making my own itemmodelgenerator for an edited model for handheld items but it didn't work because i couldn't initialize it in generateItemModels. I am not that good in programming

**Reply by theoretischer** [reply, marked as answer]
Ok i got it.
I added my own model in the same file:
```
    private final Model HANDHELD_TWO_LAYERS = item("handheld", TextureKey.LAYER0, TextureKey.LAYER1);
    private static Model item(String parent, TextureKey LAYER0, TextureKey LAYER1) {
        return new Model(Optional.of(new Identifier("item/" + parent)), Optional.empty(), LAYER0, LAYER1);
    }
```
and then i added and item by:
```
HANDHELD_TWO_LAYERS.upload(new Identifier(TestMod.MOD_ID, "item/nameofitem), TextureMap.layered(new Identifier(TestMod.MOD_ID, "item/firstlayer"), new Identifier(TestMod.MOD_ID, "item/secondlayer")), itemModelGenerator.writer);
```
